cmake/NeuronTestHelper.cmake

```diff
@@ -0,0 +1,361 @@
+# results obtained with NEURON and, optionally, CoreNEURON in different modes
+# of operation.
+#
+# 1. nrn_add_test_group(NAME name
+#                       SUBMODULE some/submodule
+#                       MODFILE_DIRECTORY mod_file_dir
+#                       OUTPUT datatype::file.ext otherdatatype::otherfile.ext
+#                       SCRIPT_PATTERNS "*.py")
+#
+#    Create a new group of integration tests with the given name. The group
+#    consists of different configurations of running logically the same
+#    simulation. The outputs of the different configurations will be compared.
+#
+#    SUBMODULE         - the name of the git submodule containing test data.
+#    MODFILE_DIRECTORY - a relative path inside the submodule that contains the
+#                        modfiles that must be compiled (using nrnivmodl) to
+#                        run the test.
+#    OUTPUT            - zero or more expressions of the form `datatype::path`
+#                        describing the output data produced by a test. The
+#                        data type must be supported by the comparison script
+#                        `compare_test_results.py`, and `path` is defined
+#                        relative to the working directory in which the test
+#                        command is run.
+#    SCRIPT_PATTERNS   - zero or more glob expressions, defined relative to the
+#                        submodule directory, matching scripts that must be
+#                        copied from the submodule to the working directory in
+#                        which the test is run.
+#
+#    The SUBMODULE, MODFILE_DIRECTORY, OUTPUT and SCRIPT_PATTERNS arguments are
+#    default values that will be inherited tests that are added to this group
+#    using nrn_add_test. They can be overriden for specific tests by passing
+#    the same keyword arguments to nrn_add_test.
+#
+# 2. nrn_add_test(GROUP group_name
+#                 NAME test_name
+#                 [REQUIRES feature1 ...]
+#                 [CONFLICTS feature1 ...]
+#                 [SUBMODULE some/submodule]
+#                 [MODFILE_DIRECTORY mod_file_dir]
+#                 [OUTPUT datatype::file.ext otherdatatype::otherfile.ext ...]
+#                 [SCRIPT_PATTERNS "*.py" ...])
+#
+#    Create a new integration test inside the given group, which must have
+#    previously been created using nrn_add_test_group. The SUBMODULE,
+#    MODFILE_DIRECTORY, OUTPUT and SCRIPT_PATTERNS arguments are optional and
+#    can be used to override the defaults defined when nrn_add_test_group is
+#    called. The REQUIRES and CONFLICTS arguments allow a test to be disabled
+#    if certain features are, or are not, available. Four features are
+#    currently supported: coreneuron, gpu, nmodl and mod_compatibility.
+#
+# 3. nrn_add_test_group_comparison(GROUP group_name
+#                                  REFERENCE_OUTPUT datatype::file.ext [...])
+#
```

Comment:
(1) yes, the comparison always runs. The dependencies just ensure that it runs after the other jobs.
(2) right now it does all the pairwise comparisons:
```
34: neuron matches neuron_mpi
34: neuron matches coreneuron_mpi
34: neuron matches coreneuron_mpi_threads
34: neuron matches reference_file
34: neuron_mpi matches coreneuron_mpi
34: neuron_mpi matches coreneuron_mpi_threads
34: neuron_mpi matches reference_file
34: coreneuron_mpi matches coreneuron_mpi_threads
34: coreneuron_mpi matches reference_file
34: coreneuron_mpi_threads matches reference_file
34/34 Test #34: external_ringtest::compare_results ..........   Passed    0.04 sec
```

It would be nice to have an ASCII-art matrix illustrating this, but for now at least the information is there.

Reply:
Looks good!